voxblox/include/voxblox/utils/neighbor_tools.h

```diff
@@ -13,246 +13,87 @@ enum Connectivity : unsigned int {
-      direction(i) = j;
-      for (int k = -1; k <= 1; k += 2) {
-        direction(next_i) = k;
-        getNeighborIndex(block_index, voxel_index, direction, &neighbor.first,
-                         &neighbor.second);
-        neighbors->emplace_back(neighbor);
-        distances->emplace_back(kSqrt2);
-        directions->emplace_back(direction);
-      }
-      direction(i) = 0;
-      direction(next_i) = 0;
-    }
-  }
-
-  if (connectivity == Connectivity::kEighteen) {
-    return;
-  }
-
-  // Distance sqrt(3) set.
-  for (int i = -1; i <= 1; i += 2) {
-    direction(0) = i;
-    for (int j = -1; j <= 1; j += 2) {
-      direction(1) = j;
-      for (int k = -1; k <= 1; k += 2) {
-        direction(2) = k;
-        getNeighborIndex(block_index, voxel_index, direction, &neighbor.first,
-                         &neighbor.second);
-        neighbors->emplace_back(neighbor);
-        distances->emplace_back(kSqrt3);
-        directions->emplace_back(direction);
-      }
-    }
-  }
-
-  if (connectivity != Connectivity::kTwentySix) {
-    LOG(FATAL) << "Unknown neighborhood connectivity: "
-               << static_cast<int>(connectivity);
-  }
-}
-
-template <typename VoxelType>
-void NeighborTools<VoxelType>::getNeighborsByGlobalIndex(
-    const GlobalIndex& global_index, Connectivity connectivity,
-    GlobalIndexVector* neighbors) const {
-  CHECK_NOTNULL(neighbors);
-  neighbors->clear();
-  neighbors->reserve(connectivity);
-
-  GlobalIndex direction;
-  direction.setZero();
-  // Distance 1 set.
-  for (unsigned int i = 0u; i < 3u; ++i) {
-    for (int j = -1; j <= 1; j += 2) {
-      direction(i) = j;
-      neighbors->emplace_back(global_index + direction);
+template <Connectivity kConnectivity>
+class Neighborhood : public NeighborhoodLookupTables {
+ public:
+  typedef Eigen::Matrix<LongIndexElement, 3, kConnectivity> IndexMatrix;
+
+  // Get the global index of all (6, 18, or 26) neighbors of the input index.
+  static void getFromGlobalIndex(const GlobalIndex& global_index,
+                                 IndexMatrix* neighbors) {
+    CHECK_NOTNULL(neighbors);
+    for (unsigned int i = 0u; i < kConnectivity; ++i) {
+      neighbors->col(i) = global_index + kLongOffsets.col(i);
     }
-    direction(i) = 0;
   }
 
-  if (connectivity == Connectivity::kSix) {
-    return;
-  }
-
-  // Distance sqrt(2) set.
-  for (unsigned int i = 0u; i < 3u; ++i) {
-    unsigned int next_i = (i + 1) % 3;
-    for (int j = -1; j <= 1; j += 2) {
-      direction(i) = j;
-      for (int k = -1; k <= 1; k += 2) {
-        direction(next_i) = k;
-        neighbors->emplace_back(global_index + direction);
+  // Get the hierarchical index (block idx, local voxel index) a voxel described
+  // by its hierarchical index and a direction. The main purpose of this
+  // function is to solve the cross-block indexing that happens when looking up
+  // neighbors at the block boundaries.
+  static void getFromHierarchicalIndexAndDirection(
```

Comment:
So Hierarchical index is defined as:
```
typedef typename HierarchicalIndexMap::value_type HierarchicalIndex;
```
In block_hash.h, so this is not from HierachicalIndex
Can we call it BlockAndVoxel index as elsewhere? I think this is much clearer and more consistent with other terminology.

Reply:
will do